File: WordPress/src/main/java/org/wordpress/android/ui/posts/services/AztecImageLoader.java
package org.wordpress.android.ui.posts.services;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.util.DisplayMetrics;

import com.android.volley.VolleyError;
import com.android.volley.toolbox.ImageLoader;

import org.wordpress.android.WordPress;
import org.wordpress.android.util.ImageUtils;
import org.wordpress.aztec.Html;

import java.io.File;

public class AztecImageLoader implements Html.ImageGetter {
    private final Drawable mLoadingInProgress;
    private Context mContext;

    public AztecImageLoader(Context context, Drawable loadingInProgressDrawable) {
        this.mContext = context;
        this.mLoadingInProgress = loadingInProgressDrawable;
    }

    @Override
    public void loadImage(final String url, final Callbacks callbacks, int maxWidth) {
        loadImage(url, callbacks, maxWidth, 0);
    }

    @Override
    public void loadImage(final String url, final Callbacks callbacks, final int maxWidth, final int minWidth) {
        final String cacheKey = url + maxWidth;
        Bitmap cachedBitmap = WordPress.getBitmapCache().get(cacheKey);
        if (cachedBitmap != null) {
            // By default, BitmapFactory.decodeFile sets the bitmap's density to the device default so, we need
            // to correctly set the input density to 160 ourselves.
            cachedBitmap.setDensity(DisplayMetrics.DENSITY_DEFAULT);
            callbacks.onImageLoaded(new BitmapDrawable(mContext.getResources(), cachedBitmap));
            return;
        }

        if (new File(url).exists()) {
            int orientation = ImageUtils.getImageOrientation(this.mContext, url);
            byte[] bytes = ImageUtils.createThumbnailFromUri(
                    mContext, Uri.parse(url), maxWidth, null, orientation);
            if (bytes != null) {
                Bitmap bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.length);
                if (bitmap != null) {
                    WordPress.getBitmapCache().putBitmap(cacheKey, bitmap);
                    bitmap.setDensity(DisplayMetrics.DENSITY_DEFAULT);
                }
                BitmapDrawable bitmapDrawable = new BitmapDrawable(mContext.getResources(), bitmap);
                callbacks.onImageLoaded(bitmapDrawable);
            } else {
                callbacks.onImageFailed();
            }
            return;
        }

        callbacks.onImageLoading(mLoadingInProgress);

        WordPress.sImageLoader.get(url, new ImageLoader.ImageListener() {
            @Override
            public void onResponse(ImageLoader.ImageContainer response, boolean isImmediate) {
                Bitmap bitmap = response.getBitmap();

                if (bitmap == null && !isImmediate) {
                    // isImmediate is true as soon as the request starts.
                    callbacks.onImageFailed();
                } else if (bitmap != null) {
                    final String cacheKey = url + maxWidth;
                    // Make sure both width ad height respect the max size for the editor
                    bitmap = ImageUtils.getScaledBitmapAtLongestSide(bitmap, maxWidth);
                    WordPress.getBitmapCache().putBitmap(cacheKey, bitmap);
                    // By default, BitmapFactory.decodeFile sets the bitmap's density to the device default so, we need
                    // to correctly set the input density to 160 ourselves.
                    bitmap.setDensity(DisplayMetrics.DENSITY_DEFAULT);
                    BitmapDrawable bitmapDrawable = new BitmapDrawable(mContext.getResources(), bitmap);
                    callbacks.onImageLoaded(bitmapDrawable);
                }
            }

            @Override
            public void onErrorResponse(VolleyError error) {
                callbacks.onImageFailed();
            }
        }, maxWidth, 0);
    }
}

File: WordPress/src/main/java/org/wordpress/android/ui/posts/services/AztecImageLoader.kt
@file:Suppress("DEPRECATION")

package org.wordpress.android.ui.posts.services

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.util.DisplayMetrics
import android.widget.ImageView.ScaleType.FIT_CENTER
import com.bumptech.glide.request.target.BaseTarget
import com.bumptech.glide.request.target.SizeReadyCallback
import com.bumptech.glide.request.transition.Transition
import org.wordpress.android.util.image.ImageManager
import org.wordpress.aztec.Html
import java.lang.ref.WeakReference

/**
 * Class which retrieves images from both local storage and remote url and informs the client about the progress
 * using Html.ImageGetter.Callbacks.
 *
 * Since Aztec doesn't allow directly setting an image into a View a custom Target which we load
 * the image into needs to be created.
 *
 * It's recommended to explicitly call clearTargets() method, so any resources may be cleared and all pending requests
 * canceled when the screen is not visible to the user anymore. Each target must be stored in a WeakReference,
 * so targets which are not used anymore don't leak.
 */
class AztecImageLoader(
    private val context: Context,
    private val imageManager: ImageManager,
    private val loadingInProgress: Drawable
) : Html.ImageGetter {
    @Suppress("DEPRECATION")
    private val targets = ArrayList<WeakReference<BaseTarget<Bitmap>>>()
    private val mRequestsInProgress = ArrayList<String>()

    override fun loadImage(url: String, callbacks: Html.ImageGetter.Callbacks, maxWidth: Int) {
        loadImage(url, callbacks, maxWidth, 0)
    }

    override fun loadImage(url: String, callbacks: Html.ImageGetter.Callbacks, maxSize: Int, minWidth: Int) {
        mRequestsInProgress.add(url)

        @Suppress("DEPRECATION") val target = object : BaseTarget<Bitmap>() {
            override fun onLoadStarted(placeholder: Drawable?) {
                callbacks.onImageLoading(loadingInProgress)
            }

            override fun onLoadFailed(errorDrawable: Drawable?) {
                callbacks.onImageFailed()
                mRequestsInProgress.remove(url)
            }

            override fun onResourceReady(
                resource: Bitmap,
                transition: Transition<in Bitmap>?
            ) {
                val result: Drawable
                // By default, BitmapFactory.decodeFile sets the bitmap's density
                // to the device default so, we need to correctly set the input density to 160 ourselves.
                resource.density = DisplayMetrics.DENSITY_DEFAULT
                val bitmapDrawable = BitmapDrawable(context.resources, resource)
                result = bitmapDrawable
                callbacks.onImageLoaded(result)
                mRequestsInProgress.remove(url)
            }

            override fun getSize(cb: SizeReadyCallback) {
                cb.onSizeReady(maxSize, maxSize)
            }

            override fun removeCallback(cb: SizeReadyCallback) {
                // since we don't store the callback, we don't need to do anything
            }
        }

        imageManager.loadAsBitmapIntoCustomTarget(context, target, url, FIT_CENTER)
        targets.add(WeakReference(target))
    }

    /**
     * Cancel any pending requests and free any resources so they may be reused.
     */
    fun clearTargets() {
        for (weakReference in targets) {
            imageManager.cancelRequest(context, weakReference.get())
        }
    }

    fun getNumberOfImagesBeingDownloaded(): Int {
        return mRequestsInProgress.count()
    }
}
